# 1. Importing Libraries and Data

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# create path
path = "/Users/charlottelin/Documents/06-2025 Instacart Basket Analysis"

In [3]:
# import products data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

In [5]:
# import wrangled orders data
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49693 rows x 5 columns]

In [7]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

# 2. Data Consistency Checks

## 2.1 Mixed-Type Data

In [8]:
# Create a test dataframe
df_test = pd.DataFrame()

In [9]:
# Create a mixed type column in df_test
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_8602/772068622.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [12]:
# Convert the mixed type column data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [13]:
df_test['mix'].dtype

dtype('O')

## 2.2 Missing Values

In [14]:
# Find missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# Create a new dataframe to subset the missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [17]:
# Find the median of the df_prods data set
df_prods['prices'].median()

7.1

In [20]:
# Check for outliers in df_prods
df_prods[df_prods['prices'] > df_prods['prices'].mean() + (2 * df_prods['prices'].std())]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [21]:
# Replace outliers with median value
df_prods = df_prods.replace({"prices":{14900.0:7.1, 99999.0:7.1}})

In [23]:
# Check df_prods stats again after imputation
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      7.682245
std    14343.717401     38.316774       5.850282      4.200159
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

#### Note: after imputing the two outliers, the standard deviation is much more reasonable at 4.2 as opposed to the 453 before.

In [18]:
# Find number of rows and columns in df_prods
df_prods.shape

(49693, 5)

In [24]:
# Create a new dataframe to exclude 'product_name'
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [25]:
df_prods_clean.shape

(49677, 5)

## 2.3 Duplicates

In [26]:
# Look for duplicates in the cleaned df_prods_clean
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [27]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [28]:
# Check number of rows in df_prods_clean
df_prods_clean.shape

(49677, 5)

In [29]:
# Drop the duplicate values in df_prods_clean
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [30]:
df_prods_clean_no_dups.shape

(49672, 5)

# 3. Task

## Step 2 - Explanatory Analysis of df_ords

In [31]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

### **Observation of the Explanatory Analysis:**
- The "unnamed: 0" column has the same values of order_id and is likely an error and might need to be dropped.
- order_id, user_id, and order_number look fine.
- The min and max values of orders_day_of_week and order_hour_of_day look as expected.
- The zero in days_since_prior_order might indicate same-day orders. Further checks should be done to see if there are missing values.

## Step 3 & 4 - Mixed Type Data & Any Fixes

In [32]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)  

/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_8602/302177923.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_8602/302177923.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_8602/302177923.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/t0/dyr0vh9d10q4v8pztlwjgtkh0000gn/T/ipykernel_8602/302177923.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].app

In [33]:
# Check the data types in df_ords
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   Unnamed: 0              int64  
 1   order_id                int64  
 2   user_id                 int64  
 3   order_number            int64  
 4   orders_day_of_week      int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(6)
memory usage: 182.7 MB


### **Observation:** no mixed-type data was found in df_ords. Therefore, no fixes are needed.

## Step 5 & 6 - Missing Values and Any Fixes

In [34]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [35]:
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

### **Observation:** The missing values in 'days_since_prior_order" is in line with the statistics observation made earlier. There are NaN present in the column. The number of 206209 requires further investigation.

In [36]:
# Further investigation of the "NaN" values by looking at the number of unique user_id
df_ords[df_ords['days_since_prior_order'].isnull()].groupby('user_id').count().describe()

Unnamed: 0  order_id  order_number  orders_day_of_week  \
count    206209.0  206209.0      206209.0            206209.0   
mean          1.0       1.0           1.0                 1.0   
std           0.0       0.0           0.0                 0.0   
min           1.0       1.0           1.0                 1.0   
25%           1.0       1.0           1.0                 1.0   
50%           1.0       1.0           1.0                 1.0   
75%           1.0       1.0           1.0                 1.0   
max           1.0       1.0           1.0                 1.0   

       order_hour_of_day  days_since_prior_order  
count           206209.0                206209.0  
mean                 1.0                     0.0  
std                  0.0                     0.0  
min                  1.0                     0.0  
25%                  1.0                     0.0  
50%                  1.0                     0.0  
75%                  1.0                     0.0  
max                  1.0                     0.0

### **Observation:** The exploratory code for user_id returned **206,209** unique users. This mataches the number of missing values in 'day_since_prior_order', which suggests that the missing value is their very first order. Therefore, **it is not necessary to remove these missing values** as each user_id would naturally create one "NaN" value. As a result, no method is needed here to adddress the missing values.

## Step 7 & 8 - Duplicates and Any Fixes

### As identified in Step 2, the "Unnamed:0" column is a duplicate and should be dropped.

In [39]:
# Verify that "unnamed:0" is a dupe
df_ords

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 7 columns]

In [40]:
#Check for other dupes
df_ords.duplicated().sum()

0

In [41]:
# Drop the "unnamed:0" column and create a new df_ords_clean dataframe
df_ords_clean = df_ords.drop('Unnamed: 0', axis=1)

In [42]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

# Step 9 - Export New Dataframes

In [43]:
# Export the checked df_prods
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

In [44]:
# Export the checked df_ords
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

### The two checked dataframes have been successfully saved in the Prepared Data folder.